### PR TITLE
Return recovery to generic thread post-PRRL action

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we perform `TransportReplicationAction` derivatives during recovery, and
these actions call their response handlers on the transport thread. This change
moves the continued execution of the recovery back onto the generic threadpool.

Relates #41536.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 1]
```diff
@@ -820,4 +827,30 @@ private void handleErrorOnSendFiles(Store store, Exception e, StoreFileMetaData[
     protected void failEngine(IOException cause) {
         shard.failShard("recovery", cause);
     }
+
+    private <Response> ActionListener<Response> withGenericExecutor(ActionListener<Response> listener) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: Use ThreadedActionListener instead?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: Thanks, that was what I was looking for.

**Review by ywelsch — APPROVED** [review]
LGTM

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java [hunk 2]
```diff
@@ -144,8 +146,10 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
                 IOUtils.closeWhileHandlingException(releaseResources, () -> wrappedListener.onFailure(e));
                 throw e;
             });
-            final Consumer<Exception> onFailure = e ->
+            final Consumer<Exception> onFailure = e -> {
+                Transports.assertNotTransportThread("failure of recovery from " + shard.routingEntry() + " to " + request.targetNode());
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: I think this should be `assert Transports.assertNotTransportThread`.

Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java]: I think this should be `assert Transports.assertNotTransportThread(...)`.

**Pull request merged**